### ISSUE TITLE
pyBRAT Error: Cannot Import name write_xml_element_with_path

### ISSUE THREAD

**Issue by joewheaton** [Riverscapes/pyBRAT#309]
From a user:

> Hello BRAT Team,
> 
> I am a grad student at UC Berkeley and I am hoping to run BRAT for the Upper Merced watershed in California (and eventually more, but starting there). I know you don’t always have time to provide support for folks on this tool, but I thought I’d email in case someone has a moment to respond.
> 

> Unfortunately my BRAT.py toolbox is broken (the red x!). I am using BRAT [3.1.00](http://3.1.0.0/) (I downloaded the source code zip file) and followed all instructions [here ](https://brat.riverscapes.xyz/Documentation/Download/ComputerSetup.html)without any issues. I am running ArcMap 10.8.1. When I “check syntax” on the broken toolbox, I get this:
> 
> Traceback (most recent call last):
>  ```
>  File "<string>", line 3, in <module>
>   File "C:\1_BRAT\pyBRAT-3.1.00\BRATProject.py", line 18, in <module>
>     from SupportingFunctions import make_folder, make_layer, get_execute_error_code, write_xml_element_with_path, find_available_num_prefix
> ImportError: cannot import name write_xml_element_with_path
> ```
> 
> I’m not very familiar with Python, but from what I can find on stack exchange, that type of error might indicate some sort of circular dependency? But I’m not sure. I also combed through your GitHub issues, and while [this ](https://github.com/Riverscapes/pyBRAT/issues/235)post is similar, it didn't quite answer my question (but did suggest I might need a Python wizard). Any advice would be appreciated!
> 
> Also, if you rather I post this on GitHub please let me know– it seemed like the GitHub hadn’t been active for about a year, so I figured I’d email first! If I don’t hear back in a bit I’ll post there too.

**Comment by joewheaton** [comment]
> I am hoping to run BRAT for the Upper Merced watershed in California (and eventually more, but starting there). I know you don’t always have time to provide support for folks on this tool, but I thought I’d email in case someone has a moment to respond.

So, if you sign up for an account at http://data.riverscapes.xyz/, and email when you have the email address you used, there are some pre-existing pyBRAT runs in a TNC_BRAT Program you can download.  It looks like that one may not have gotten uploaded for some reason (1804008), but I did find it here https://usu.app.box.com/s/nngn8j4l92wtleob47wzdaoxvb9vdq52/folder/81951491843,  that got run as part of that [TNC BRAT](http://brat.riverscapes.xyz/Regions/USA/TNC_SierraNevada/) project a few years back. Somewhere, TNC has a viewer of this too. 


> Unfortunately my BRAT.py toolbox is broken (the red x!). I am using BRAT [3.1.00](http://3.1.0.0/) (I downloaded the source code zip file) and followed all instructions [here ](https://brat.riverscapes.xyz/Documentation/Download/ComputerSetup.html)without any issues. I am running ArcMap 10.8.1. When I “check syntax” on the broken toolbox, I get this:
> 
> ```
> Traceback (most recent call last):
>   File "<string>", line 3, in <module>
>   File "C:\1_BRAT\pyBRAT-3.1.00\BRATProject.py", line 18, in <module>
>     from SupportingFunctions import make_folder, make_layer, get_execute_error_code, write_xml_element_with_path, find_available_num_prefix
> ImportError: cannot import name write_xml_element_with_path
> ```


​Python dependencies are a nightmare to troubleshoot and support. If you don't like the above run for some reason or want to modify it, we can chat. We have gotten the price-tag down to $400 a HUC 8 to run these, but we have done that by refactoring this to sqlBRAT (https://tools.riverscapes.xyz/brat/) and deploying it in the cloud on AWS. It is not something that users can run on their own very easily (it is all open source, but you need to know AWS inside and out, which very few people do).  We realized that there was a much bigger audience of people wanting to use the model outputs as a tool, then the audience of folks wanting to run it themselves.  Someday, we'd like to support researchers with some basic GIS to run their own runs, but the truth is that is about a $200K to $350K development task to move it from [research grade](https://riverscapes.xyz/Tools/discrimination.html#tool-grade), fragile ArcPy scripts to a [professional grade](https://riverscapes.xyz/Tools/discrimination.html#tool-grade) desktop GIS tool that works. However, we made a production grade tool that is truly scalable and we can run for the entire country. 

All that said, the **error message seems like it is just struggling to write the XML to that path.** These things are super fiddly and it might not like the `.` periods in the path you have `C:\1_BRAT\pyBRAT-3.1.00\BRATProject.py`. You could try changing the location of BRATProject.py and all its dependencies to a folder called  `C:\1_BRAT\pyBRAT3_1\BRATProject.py` and see what happens. That will probably just kick the can down the road to the next error message.  The unfortunate truth is it typically costs us more in staff time to hold someone's hand then just run the model... but maybe we get lucky this time?  


> I’m not very familiar with Python, but from what I can find on stack exchange, that type of error might indicate some sort of circular dependency? But I’m not sure. I also combed through your GitHub issues, and while [this ](https://github.com/Riverscapes/pyBRAT/issues/235)post is similar, it didn't quite answer my question (but did suggest I might need a Python wizard). Any advice would be appreciated!

>Also, if you rather I post this on GitHub please let me know– it seemed like the GitHub hadn’t been active for about a year, so I figured I’d email first! If I don’t hear back in a bit I’ll post there too.

Yes, we are no longer supporting this Python version, so that's why there isn't much traffic. I'll still port my response here as others will no doubt have similar troubles. 

Let us know if any of above is of interest or works.

**Comment by jessiemoravek** [comment]
Thanks for your helpful response! 

1) We did look into the previous TNC BRAT run for the state of California, but my understanding is that it used a 2014 LandFire dataset as the vegetation layer input. There have been a lot of catastrophic fires in California since 2014, so we thought it would be more appropriate to re-run the model using 2020 LandFire data. But let me know if you have any thoughts on how big fires might influence model outputs!

2) Thanks for the support, even if this is an unsupported Python version of BRAT at this point! We tried your suggestion about changing “.” to “_”, but unfortunately no luck! If we are ever able to troubleshoot that dependency, I’ll post the solution here.

**Comment by joewheaton** [comment]
> Thanks for your helpful response!
> 
No problem.

> 1. We did look into the previous TNC BRAT run for the state of California, but my understanding is that it used a 2014 LandFire dataset as the vegetation layer input. There have been a lot of catastrophic fires in California since 2014, so we thought it would be more appropriate to re-run the model using 2020 LandFire data. But let me know if you have any thoughts on how big fires might influence model outputs!

Great point. Yes, we can do re-runs with new platform (using old calibration data for vegetation, but newer LANDFIRE releases).  We do have a project with Emily Fairfax and USFWS exploring some of changes post-fire, as well as looking at fire modelling to see if more beaver dam activity was present if that might have changed fire behavior on some of the recent megafires.  

We're hoping to get to the point eventually where 

> 2. Thanks for the support, even if this is an unsupported Python version of BRAT at this point! We tried your suggestion about changing “.” to “_”, but unfortunately no luck! If we are ever able to troubleshoot that dependency, I’ll post the solution here.

I will leave this issue open for now in case @jtgilbert or @Jdgilby have any suggestions on getting pyBRAT going. 

Also, @philipbaileynar and @jtgilbert do you have any ETA on when some basic documentation might come out on how to do "local" runs of [sqlBRAT](http://tools.riverscapes.xyz/brat) not using cybercastor on AWS? I know it is a pain, and very few are likely to engage, but some researchers might go there. @jessiemoravek this is something that is likely easier to set up to run on Linux or Mac then PC, but can still be done. We are also exploring ways custom runs can be request (e.g. with specific DEMs differing from NED → like 3DEP; or with specific LANDFIRE dates, or with custom vegetation inputs, or custom extents). Those are likely a few years out, but I bring it up here in case @philipbaileynar wants to comment.

**Comment by jessiemoravek** [comment]
Hi! Thanks for the info about post-fire modelling! I'm excited to see the results of that work. I also look forward to seeing sqlBRAT documentation down the road! 

In the meantime, we found a workaround to our pyBRAT toolbox problems. The pyBRAT toolbox works when accessed through the ArcCatalog, even though it's broken when accessing through ArcMap. We were able to successfully run a BRAT model through catalog and used ArcMap to view the results. Big thanks to Eric in the Berkeley GIF lab for figuring that one out.
